File: other-lang/ruby/reversecomplement.rb
# The Computer Language Benchmarks Game
# http://shootout.alioth.debian.org/
# reverse-complement Ruby 1.9 #2
#
# Contributed by Peter Bjarke Olsen
# Modified by Doug King
# Modified by Joseph LaFata
# Slightly modified by Leo Osvald

seq=""

def revcomp(seq)
  seq.reverse!.tr!('wsatugcyrkmbdhvnATUGCYRKMBDHVN','WSTAACGRYMKVHDBNTAACGRYMKVHDBN')
  stringlen=seq.length-1
  0.step(stringlen,60) {|x| print seq[x,60] , "\n"}
end

File.open(ARGV[0]).each do |line|
  if line.include? '>'
    if !seq.empty?
      revcomp(seq)
      seq=""
    end
    puts line
  else
    line.chomp!
    seq << line
  end
end
revcomp(seq)

File: other-lang/python/2/reversecomplement.py
# The Computer Language Benchmarks Game

# http://shootout.alioth.debian.org/
# reverse-complement Python 3 #4

#

# contributed by Jacob Lee, Steven Bethard, et al
# 2to3, fixed by Daniele Varrazzo
# modified by Daniel Nanz
# slightly modified by Leo Osvald

import string
import sys

def show(seq, table=string.maketrans(b'ACBDGHKMNSRUTWVYacbdghkmnsrutwvy',
                                     b'TGVHCDMKNSYAAWBRTGVHCDMKNSYAAWBR'),
         write=sys.stdout.write, nl=b'\n'):
    
    [header, s] = seq.split(nl, 1)
    s = s.translate(table, nl)[: : -1]
    
    write(b'>' + header + nl)
    for i in range(0, len(s), 60):
        write(s[i : i + 60] + nl)



def main():
    f = open(sys.argv[1], 'r')
    #sys.stdin = sys.stdin.detach()
    seqs = b''.join([line for line in f]).split(b'>')[1 : ]
    
    for seq in seqs:
        show(seq)        


main()
